uniffi_bindgen/src/lib.rs

```diff
@@ -110,53 +110,49 @@ pub mod scaffolding;
 pub use interface::ComponentInterface;
 use scaffolding::RustScaffolding;
 
-/// Trait for bindings configuration.  Each bindings language defines one of these.
-///
-/// BindingsConfigs are initially loaded from `uniffi.toml` file.  Then the trait methods are used
-/// to fill in missing values.
-pub trait BindingsConfig: DeserializeOwned {
-    /// Update missing values using the `ComponentInterface`
-    fn update_from_ci(&mut self, ci: &ComponentInterface);
-
-    /// Update missing values using the dylib file for the main crate, when in library mode.
-    ///
-    /// cdylib_name will be the library filename without the leading `lib` and trailing extension
-    fn update_from_cdylib_name(&mut self, cdylib_name: &str);
-
-    /// Update missing values from config instances from dependent crates
-    ///
-    /// config_map maps crate names to config instances. This is mostly used to set up external
-    /// types.
-    fn update_from_dependency_configs(&mut self, _config_map: HashMap<&str, &Self>) {}
-}
-
 /// A trait representing a UniFFI Binding Generator
 ///
 /// External crates that implement binding generators, should implement this type
 /// and call the [`generate_external_bindings`] using a type that implements this trait.
 pub trait BindingGenerator: Sized {
     /// Handles configuring the bindings
-    type Config: BindingsConfig;
+    type Config;
 
     /// Creates a new config.
     fn new_config(&self, root_toml: &toml::Value) -> Result<Self::Config>;
 
+    /// Update the various config items in preparation to write one or more of them.
+    ///
+    /// # Arguments
+    /// - `cdylib`: The name of the cdylib file, if known.
+    /// - `library_path`: The name of library used to extract the symbols.
+    /// - `components`: A mutable array of [`Component`]s to be updated.
+    fn update_component_configs(
+        &self,
+        cdylib: &Option<&str>,
+        library_path: Option<&Utf8Path>,
+        components: &mut Vec<Component<Self::Config>>,
+    ) -> Result<()>;
+
     /// Writes the bindings to the output directory
     ///
     /// # Arguments
-    /// - `ci`: A [`ComponentInterface`] representing the interface
-    /// - `config`: An instance of the [`BindingsConfig`] associated with this type
+    /// - `components`: An array of [`Component`]s representing the items to be generated.
     /// - `out_dir`: The path to where the binding generator should write the output bindings
     fn write_bindings(
         &self,
-        ci: &ComponentInterface,
-        config: &Self::Config,
+        components: &[Component<Self::Config>],
         out_dir: &Utf8Path,
         try_format_code: bool,
     ) -> Result<()>;
+}
```

Comment:
Would it be simpler to just have a single function that does the work of all of the `BindingGenerator` methods?  This means we don't need the `Config` type anymore, that data would just be a local variables in that method.  I think this would actually simplify the generator code for the builtin bindings.

Reply:
So a single function that takes mutable references, then does both the updating of the configs *and* writing the bindings? I did think about that and left it as it was because I kinda thought the separation of mutability and the generation seemed nice - it's not that I thought it was a bad idea, I just left it alone because I could see a reason to split them. But yeah, I am inclined to agree that maybe that's even better.

So to be clear, you'd suggest having that function take a vec of roughly (`CI`, `toml::Value`), then convert to Config objects inside that one method? I think this function would need to return the `Config` objects though - eg, [this swift code](https://github.com/mozilla/uniffi-rs/blob/main/uniffi_bindgen/src/bindings/swift/test.rs#L140) need to know what `Config` is.

(Maybe I'll just take smaller steps - stick with `Config` for now then see what people think once the "new world" is a little clearer)

Reply:
Assuming you meant what I thought you meant, another slight problem here is that we filter the CIs - ie, the list of CIs we pass to `update_component_configs` might have more items then what we pass to `write_bindings`.